File: LeetCode_Algorithm/25(reverse_nodes_in_k_groups)(linked_list)/25.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from email import header
import tarfile
from typing import Optional
class Solution:
    def makeLinkedList(self, nodes, idx):
        if idx >= len(nodes):
            return None
        root = ListNode(nodes[idx])
        root.next = self.makeLinkedList(nodes, idx+1)
        return root

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # 87ms 33%
        cnt = 0; node = head
        while node and cnt < k:
            node = node.next
            cnt += 1
        if cnt < k: return head
        new_head, prev = self.reverseLinkedList(head, cnt)
        head.next = self.reverseKGroup(new_head, k)
        return prev

    def reverseLinkedList(self, head, cnt):
        prev, curr, next = None, head, head
        while cnt > 0:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
            cnt -= 1
        return (curr, prev)

    def print(self, root):
        while root:
            print(root.val)
            root = root.next
        return
        
root = Solution().makeLinkedList([1,2,3,4,5], 0)
root = Solution().reverseKGroup(root, 2)
Solution().print(root)

File: LeetCode_Algorithm/25(reverse_nodes_in_k_groups)(linked_list)/25.js
var reverseKGroup = function(head, k) {
  // 127ms 36%
  let cnt = 0; let node = head
  while (node && cnt < k) {
    node = node.next
    cnt++
  }
  if (cnt < k) {return head}
  temp = reverseLinkdedList(head, k)
  let newHead = temp[0]; let prev = temp[1]
  head.next = reverseKGroup(newHead, k)
  return prev
};

function reverseLinkdedList(head, cnt) {
  let prev = null; let curr = head; let next = head
  while (cnt > 0) {
    next = curr.next
    curr.next = prev
    prev = curr
    curr = next
    cnt -= 1
  }
  return [curr, prev]
}


